src/Sylius/Bundle/AddressingBundle/Repository/ZoneRepository.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\AddressingBundle\Repository;
+
+use Doctrine\ORM\QueryBuilder;
+use Sylius\Bundle\ResourceBundle\Doctrine\ORM\EntityRepository;
+use Sylius\Component\Addressing\Model\AddressInterface;
+use Sylius\Component\Addressing\Model\Scope;
+use Sylius\Component\Addressing\Model\ZoneInterface;
+use Sylius\Component\Addressing\Repository\ZoneRepositoryInterface;
+
+/**
+ * @implements ZoneRepositoryInterface<ZoneInterface>
+ */
+class ZoneRepository extends EntityRepository implements ZoneRepositoryInterface
+{
+    public function findOneByAddress(AddressInterface $address, ?string $scope = null): ?ZoneInterface
+    {
+        $query = $this->createFindByAddressQueryBuilder($address, $scope);
+
+        $query
+            ->addSelect('(CASE
+                    WHEN z.type = \'province\' THEN 1
+                    WHEN z.type = \'country\' THEN 2
+                    WHEN z.type = \'zone\' THEN 3
+                    ELSE 4
+                END) AS HIDDEN sort_order')
+            ->orderBy('sort_order', 'ASC')
+            ->setMaxResults(1)
```

Comment:
I'm not convinced on this technique.

If query is constructed to return single result and there are more than one results, then we should get an `Exception` instead of returning first available result.

It's sort of a `LogicException` case, which should lead directly to a fix in the code by developer. (query itself or entity mappings).

Can we have more opinion on this @Sylius/core-team (Does it work to tag whole team? :D)

Reply:
Reproducing the previous logic with a query, this is what we want to achieve, to get one zone with the highest priority or just the first one when the priorities are the same, even if more than one zone matches the address. Perhaps it would be good to rename the method to make it clearer, sth like: `findOneWithHighestPriorityByAddress` or move the logic of choosing the one zone to the ZoneMatcher 🤔